File: LeetCode/C++/200-299/215.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
    int quickSelect(vector<int>& nums, int k, int left, int right)
    {
        int pivot = nums[left];
        int i = left;
        int j = right;
        while(i <= j)
        {
            while(nums[i] > pivot)
            {
                i++;
            }

            while(nums[j] < pivot)
            {
                j--;
            }

            if(i <= j)
            {
                swap(nums, i, j);
                i++;
                j--;
            }
        }

        if(left < j && k <= j)
            return quickSelect(nums, k, left, j);

        if(i < right && k >= i)
            return quickSelect(nums, k, i, right);

        return nums[k];
    }

    void swap(vector<int>& nums, int i, int j)
    {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSelect(nums, k - 1, 0, nums.size() - 1);
    }
};

int main()
{
    vector<int>nums = {4,1,6,5,9,2};
    Solution s;
    cout << s.findKthLargest(nums, 2) << endl;
    return 0;
}

File: LeetCode/C#/200-299/215.cs
using System;

namespace LeetCode215
{
    public class Solution {
        private void Swap(int[] nums, int i, int j)
        {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
        
        private int QuickSelect(int[] nums, int k, int left , int right)
        {
            int i = left;
            int j = right;
            int pivot = nums[left];
            int len = nums.Length;
            while(i <= j)
            {
                while(nums[i] > pivot)
                {
                    i++;
                }
                
                while(nums[j] < pivot)
                {
                    j--;
                }
                
                if(i <= j)
                {
                    Swap(nums, i, j);
                    i++;
                    j--;
                }
            }
            
            if(left < j && k <= j)
                return QuickSelect(nums, k, left, j);
            
            if(right > i && k >= i)
                return QuickSelect(nums, k, i, right);
            
            return nums[k];
        }
        
        public int FindKthLargest(int[] nums, int k) {
            return QuickSelect(nums, k - 1 , 0, nums.Length - 1);
        }
        
        
    }
    class Program
    {
        static void Main(string[] args)
        {
            Solution s = new Solution();
            int[] nums = {4,1,6,5,9,2};
            Console.WriteLine(s.FindKthLargest(nums, 2));
        }
    }
}
